contracts/asset/src/execute/reservation.rs

```diff
@@ -0,0 +1,97 @@
+use cosmwasm_std::{CustomMsg, DepsMut, Env, MessageInfo, Response};
+use cw721::{Expiration, traits::Cw721State};
+
+use crate::{
+    error::ContractError,
+    state::{AssetConfig, Reserve},
+};
+
+use super::permissions::check_can_list;
+
+pub fn reserve<TNftExtension, TCustomResponseMsg>(
+    deps: DepsMut,
+    env: &Env,
+    info: &MessageInfo,
+    id: String,
+    reservation: Reserve,
+) -> Result<Response<TCustomResponseMsg>, ContractError>
+where
+    TNftExtension: Cw721State,
+    TCustomResponseMsg: CustomMsg,
+{
+    let asset_config = AssetConfig::<TNftExtension>::default();
+
+    let mut listing = asset_config
+        .listings
+        .may_load(deps.storage, &id)?
+        .ok_or_else(|| ContractError::ListingNotFound { id: id.clone() })?;
+
+    // only the ones who can list can reserve
+    let nft_info = asset_config.cw721_config.nft_info.load(deps.storage, &id)?;
+    check_can_list(deps.as_ref(), env, info.sender.as_ref(), &nft_info)?;
+
+    if let Some(reserved) = &listing.reserved {
+        if !Expiration::AtTime(reserved.reserved_until).is_expired(&env.block) {
+            return Err(ContractError::ReservedAsset { id: id.clone() });
+        }
+    }
+
+    listing.reserved = Some(Reserve {
+        reserver: reservation.reserver.clone(),
+        reserved_until: reservation.reserved_until,
```

Comment:
Is there any bound on this value? Or can I just reserve something for 10,000,000 years?

Reply:
there is a time lock plugin that sets the maximum duration an asset can be reserved for and is enforced upon reservation but if the time lock plugin is not set, the asset can be reserved indefinitely until unreserved by the owner or approvals

Reply:
@ash-burnt